### PR TITLE
[trajoptlib] Add max friction force constraint

### PR DESCRIPTION
**Pull request by oh-yes-0-fps**
It doesn't appear to have a substantial impact of trajectories but it never urts to have.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -3,13 +3,15 @@
...
     "files.associations": {
+        "*.py": "python",
         "*.traj": "json",
         "vector": "cpp",
         "array": "cpp",
         "deque": "cpp",
         "map": "cpp",
         "unordered_map": "cpp",
-        "xstring": "cpp"
+        "xstring": "cpp",
```
Comment (calcmogul) [inline .vscode/settings.json:13]: This seems like a compiler internal header that shouldn't be here.

File: trajoptlib/src/SwerveTrajectoryGenerator.cpp [hunk 2]
```diff
@@ -245,9 +245,15 @@ SwerveTrajectoryGenerator::SwerveTrajectoryGenerator(
                             Fy.at(index).at(moduleIndex)};
       double maxForce =
           path.drivetrain.wheelMaxTorque / path.drivetrain.wheelRadius;
+      double mgCof = path.drivetrain.mass * 9.8 * path.drivetrain.wheelCof;
```
Comment (calcmogul) [inline trajoptlib/src/SwerveTrajectoryGenerator.cpp:248]:
```suggestion
      double maxFrictionForce = path.drivetrain.mass * 9.8 * path.drivetrain.wheelCoF;
```

File: trajoptlib/include/trajopt/SwerveTrajectoryGenerator.hpp [hunk 3]
```diff
@@ -36,6 +36,9 @@ struct TRAJOPT_DLLEXPORT SwerveDrivetrain {
   /// Maximum torque applied to wheel (N−m).
   double wheelMaxTorque;
 
+  /// The Coefficient of Friction (Cof) for the wheels.
+  double wheelCof;
```
Comment (calcmogul) [inline trajoptlib/include/trajopt/SwerveTrajectoryGenerator.hpp:40]:
```suggestion
  /// The Coefficient of Friction (CoF) for the wheels.
  double wheelCoF;
```

File: trajoptlib/src/SwerveTrajectoryGenerator.cpp [hunk 4]
```diff
@@ -245,9 +245,15 @@ SwerveTrajectoryGenerator::SwerveTrajectoryGenerator(
...
 
+      double forceToConstrainTo = std::min(maxForce, mgCof);
+
+      // |F|₂ ≤ mg * cof
+      // or
       // |F|₂² ≤ Fₘₐₓ²
```
Comment (calcmogul) [inline trajoptlib/src/SwerveTrajectoryGenerator.cpp:254]:
```suggestion
      // |F|₂² ≤ μmg
      // or
      // |F|₂² ≤ Fₘₐₓ²
```

**Comment by oh-yes-0-fps** [comment]
This also appears to have no substantial impact on performance when using a typical COF, tested against beta-6

File: trajoptlib/src/SwerveTrajectoryGenerator.cpp [hunk 5]
```diff
@@ -245,9 +245,15 @@ SwerveTrajectoryGenerator::SwerveTrajectoryGenerator(
                             Fy.at(index).at(moduleIndex)};
       double maxForce =
           path.drivetrain.wheelMaxTorque / path.drivetrain.wheelRadius;
+      double mgCof = path.drivetrain.mass * 9.8 * path.drivetrain.wheelCof;
 
+      double forceToConstrainTo = std::min(maxForce, mgCof);
```
Comment (calcmogul) [inline trajoptlib/src/SwerveTrajectoryGenerator.cpp:250]:
```suggestion
      double maxForce = std::min(maxWheelForce, maxFrictionForce);
```

File: trajoptlib/src/SwerveTrajectoryGenerator.cpp [hunk 6]
```diff
@@ -245,9 +245,15 @@ SwerveTrajectoryGenerator::SwerveTrajectoryGenerator(
                             Fy.at(index).at(moduleIndex)};
       double maxForce =
```
Comment (calcmogul) [inline trajoptlib/src/SwerveTrajectoryGenerator.cpp:246]:
```suggestion
      double maxWheelForce =
```

File: trajoptlib/src/SwerveTrajectoryGenerator.cpp [hunk 7]
```diff
@@ -245,9 +245,15 @@ SwerveTrajectoryGenerator::SwerveTrajectoryGenerator(
...
+      double mgCof = path.drivetrain.mass * 9.8 * path.drivetrain.wheelCof;
 
+      double forceToConstrainTo = std::min(maxForce, mgCof);
+
+      // |F|₂ ≤ mg * cof
+      // or
       // |F|₂² ≤ Fₘₐₓ²
-      problem.SubjectTo(moduleF.SquaredNorm() <= maxForce * maxForce);
+      problem.SubjectTo(moduleF.SquaredNorm() <=
+                        forceToConstrainTo * forceToConstrainTo);
```
Comment (calcmogul) [inline trajoptlib/src/SwerveTrajectoryGenerator.cpp:256]:
```suggestion
                        maxForce * maxForce);
```

**Review by calcmogul — CHANGES_REQUESTED** [review]

**Review by calcmogul — APPROVED** [review]

**Pull request merged**